### ISSUE TITLE
S3 GetObject Memory Leak when multiple calls made in parallel

### ISSUE THREAD

**Issue by jfstephe** [aws/aws-sdk-js#4017]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [ x] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/)
- [ x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js) for answers
- [x ] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js/issues) and didn't find any solution
- [ x] This is an issue with version 2.x of the SDK

**Describe the bug**
When obtaining a 30Mb file from S3 in node, I'm seeing the memory to increase and not be reclaimed. I've tried running the GC manually in a dev context and am reasonably confident this is an issue.

At the bottom of this issue is a nodejs script to reproduce the issue. It uses 2 different methods to obtain the data (AWS API and fetch) and 2 different tests (in-parallel vs in-series). Each should be commented in/out as appropriate.

The index.js file obtains the file from S3, 12 times, either a) in parallel, or b) in series. These tests were performed using both the AWS API and the built in 'fetch' command to access the S3 data directly for a comparison.

When not running any tests the baseline memory usage was approx 53Mb.

Using the AWS API: 
IN-PARALLEL TEST - AWS api: **310Mb** after completion
IN-SERIES TEST - AWS api: **145Mb** after completion

Using FETCH: 
IN-PARALLEL TEST - **73Mb** after completion
IN-SERIES TEST - **71Mb** after completion

The first thing to say is that the AWS API is leaving a lot more in use (not sure why). The other more serious problem for me is that the in parallel tests leave double the memory after all the work has been completed.

**Is the issue in the browser/Node.js?**
Node.js

**If on Node.js, are you running this on AWS Lambda?**
No

**Details of the browser/Node.js version**
v16.13.0

**SDK version number**
2.1000.0
* NOTE: I also briefly tested this with v3 and think I saw similar issues but reverted as the AWSError.code doesn't seem to have an equivalent in S3 and the documentation in S3 is awful.*

**To Reproduce (observed behavior)**

**Expected behavior**
Memory to remain consistent.

**Screenshots**
N/A

**Additional context**

```js

const AWS = require("aws-sdk");
const https = require('https')

function getFile() {
  let promise;
  ///////////////////////////////////
  // AWS API
  const s3 = new AWS.S3({
    apiVersion: '2006-03-01',
    endpoint: 's3.amazonaws.com',
    region: 'eu-west-1'
  });

  const awsGetObjectRequest = {
    Bucket: 'SOME_BUCKET',
    Key: `SOME_KEY`
  };
  console.log(`S3FileProvider.getFile - ${awsGetObjectRequest.Bucket}:${awsGetObjectRequest.Key}`);
  promise = s3.getObject(awsGetObjectRequest).promise();

  ///////////////////////////////////
  // FETCH
 /* promise = new Promise((resolve, reject) => {
    const url = "https://s3-eu-west-1.amazonaws.com/LARGE_FILE_LOCATION?AWSAccessKeyId=SOME_KEY&Expires=SOME_TIME&Signature=BLAH";
    https.get(url, res => {
      let data = '';
      res.on('data', chunk => {
        data += chunk;
      });
      res.on('end', () => {
        resolve({ Body: data});
      })
    }).on('error', err => {
      reject(err);
    }).end();
  });*/
  ///////////////////////////////////

  return promise;
}

function runTest() {
  console.log('runTest...')
  let promises = [];

  // IN-PARALLEL TEST
  // AWS api: 310Mb after completion
  // Fetch: 73Mb after completion
  const limit = 12;
  for (let count = 0; count < limit; count++) {
    promises.push(getFile().then((rawS3File) => {
      const bodyAsString = rawS3File.Body.toString();
      console.log(`${count}/${limit}`);
    }));
  }

  /////////////////
  // IN-SERIES TEST
  // AWS api: 145Mb after completion
  // Fetch: 71Mb after completion
 /* promises.push(
  getFile().then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('1')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('2')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('3')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('4')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('5')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('6')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('7')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('8')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('9')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('10')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('11')
    return getFile();
  }).then((rawS3File) => {
    const bodyAsString = rawS3File.Body.toString();
    console.log('12');
    return bodyAsString;
  }));*/
  ////////////////////////////////

  console.log('waiting...')
  Promise.all(promises).then(() => {
    console.log('done')
  });
}

// Baseline without calling 'runTest()' was 53Mb
console.log('Starting in 5 seconds...')
setTimeout(runTest, 5000);

const pollTime = 1000000;
let interval = setInterval(() => {}, pollTime);

```

**Comment by ajredniwja** [comment]
Hey @jfstephe thanks for opening this issue and apologies as it fell out of queue. I will test the same and add my results but I believe it can be more of an enhancement than a bug with the SDK.

**Comment by ajredniwja** [comment]
Would also try to see V3 results

**Comment by tylerschoppe** [comment]
Seeing similar behavior using V3.

**Comment by jfstephe** [comment]
@ajredniwja any joy on this?

(I see a fair few posts about high CPU, which I'm currently investigating another issue we've seen which I think may be related to S3).

Would be great to have an idea of any known issues and fix ETAs on these non-functional requirements of small CPU and memory footprint.

**Comment by jfstephe** [comment]
https://github.com/aws/aws-sdk-js/issues/3128 talks about a workaround on v3 (which had awful documentation last time I looked) but also sounds like it has some bugs in this area. Seems like a memory leak minefield out there. Any guidance/fixes would be gratefully received.

This issue is taking out production EC2 instances ATM...

**Comment by aBurmeseDev** [comment]
Hi everyone - checking in here. Is it still ongoing issue? This is likely due to the way Node.js handles buffer objects and their underlying memory allocation. When you fetch a file from S3 using the AWS SDK or the https module, the data is stored in a buffer object. However, Node.js does not automatically reclaim the memory occupied by these buffer objects unless they are explicitly garbage collected.

The reason why the parallel tests consume more memory is that all the file fetches are happening simultaneously and the buffer objects are being created and stored in memory at the same time. In the series test, each file fetch happens one after the other allowing the garbage collector to reclaim the memory used by the previous buffer object before allocating memory for the next one.
